File: marg/src/marg/ast/other/Statement.java
package marg.ast.other;

import marg.ast.statement.*;
import marg.debug.Console;
import marg.exception.ParseException;
import marg.debug.Debug;
import marg.lang.type.IType;
import marg.parser.Environment;
import marg.ast.ASTree;
import marg.token.TokenSet;

import java.io.EOFException;

public class Statement extends ASTree {
    ASTree child;
    public Statement(TokenSet ls) throws ParseException {
        try {
            child =
              ls.is("while")  ? new While(ls)
            : ls.is("if")     ? new If(ls)
            : ls.is("unless") ? new Unless(ls)
            : ls.is("echo")   ? new Echo(ls)
            : ls.isName()     ? new Assign(ls)
            :                   new Expr(ls)
            ;
        }
        catch (EOFException e) {
            return;
        }
        catch (Exception e) {
            Console.out.println("Unexpected error.");
            e.printStackTrace(Console.out);
        }
    }

    @Override
    public IType eval(int k, Environment e) {
        Debug.log(k, "Statement");
        return child.eval(k + 1, e);
    }
}

File: marg/src/marg/ast/other/Statement.scala
package marg.ast.other

import java.io.EOFException
import marg.ast.base.ASTree
import marg.ast.statement._
import marg.parser.Env
import marg.token.TokenSet


class Statement extends ASTree {
  private var child: ASTree = null

  def this(ls: TokenSet) {
    this()
    try {
      child = ls.get.String match {
        case "var"    => new SDefine(ls)
        case "while"  => new SWhile(ls)
        case "if"     => new SIf(ls)
        case "unless" => new SUnless(ls)
        case "echo"   => new SEcho(ls)
        case _ =>
          if (ls.isName) new SAssign(ls)
          else new Expr(ls)
      }
    }
    catch {
      case e: EOFException =>
    }
  }

  override def eval(e: Env) = child.eval(e)
}
